x-pack/platform/plugins/shared/ml/public/application/model_management/trained_models_service.ts

```diff
@@ -0,0 +1,633 @@
+
+    return items.map((item) => {
+      const previous = existingItems.find((m) => m.model_id === item.model_id);
+      const merged = {
+        ...item,
+        spaces: spaces[item.model_id] ?? [],
+      };
+
+      if (!previous || !isBaseNLPModelItem(previous) || !isBaseNLPModelItem(item)) {
+        return merged;
+      }
+
+      // Preserve "DOWNLOADING" state and the accompanying progress if still in progress
+      if (previous.state === MODEL_STATE.DOWNLOADING) {
+        return {
+          ...merged,
+          state: previous.state,
+          downloadState: previous.downloadState,
+        };
+      }
+
+      // If was "STARTING" and there's still a scheduled deployment, keep it in "STARTING"
+      if (
+        previous.state === MODEL_STATE.STARTING &&
+        this.scheduledDeployments.some((d) => d.modelId === item.model_id) &&
+        item.state !== MODEL_STATE.STARTED
+      ) {
+        return {
+          ...merged,
+          state: previous.state,
+        };
+      }
+
+      return merged;
+    });
+  }
+
+  private setupFetchingSubscription() {
+    this.subscription.add(
+      this._reloadSubject$
+        .pipe(
+          tap(() => this._isLoading$.next(true)),
+          switchMap(() => {
+            const modelsList$ = from(this.trainedModelsApiService.getTrainedModelsList()).pipe(
+              catchError((error) => {
+                this.displayErrorToast?.(
+                  error,
+                  i18n.translate('xpack.ml.trainedModels.modelsList.fetchFailedErrorMessage', {
+                    defaultMessage: 'Error loading trained models',
+                  })
+                );
+                return of([] as TrainedModelUIItem[]);
+              })
+            );
+
+            const spaces$ = this.canManageSpacesAndSavedObjects
+              ? from(this.savedObjectsApiService.trainedModelsSpaces()).pipe(
+                  catchError(() => of({})),
+                  map(
+                    (spaces) =>
+                      ('trainedModels' in spaces ? spaces.trainedModels : {}) as Record<
+                        string,
+                        string[]
+                      >
+                  )
+                )
+              : of({} as Record<string, string[]>);
+
+            return forkJoin([modelsList$, spaces$]).pipe(
+              finalize(() => this._isLoading$.next(false))
+            );
+          })
+        )
+        .subscribe(([items, spaces]) => {
+          const updatedItems = this.mergeModelItems(items, spaces);
+          this._modelItems$.next(updatedItems);
+          this.startDownloadStatusPolling();
+        })
+    );
+  }
+
+  private setupDeploymentSubscription() {
+    this.subscription.add(
+      this._scheduledDeployments$
+        .pipe(
+          distinctUntilChanged((prev, curr) => isEqual(prev, curr)),
+          mergeMap((deployments) =>
+            from(deployments).pipe(mergeMap((deployment) => this.handleDeployment(deployment)))
```

Comment:
nit: slightly easier to read:
```suggestion
            merge(...deployments.map(deployment => this.handleDeployment(deployment)))
```

Reply:
Done in: https://github.com/elastic/kibana/pull/205699/commits/302723c05d181677e002312f1d1bcaf6e8b22a03